torchdynamo/eval_frame.py

```diff
@@ -511,6 +585,59 @@ def output(self, target, args, kwargs):
         graph,
     ).transform()
 
+    if guard_args:
+        id_to_arg_mapping = dict()
+        for i in range(len(args)):
+            a = args[i]
+            id_to_arg_mapping[id(a)] = (i, a)
+        placeholders = [n for n in new_graph.graph.nodes if n.op == "placeholder"]
+        if len(placeholders) > 0:
+            for guard in out_guards:
+                if "TENSOR_MATCH" in guard.guard_types:
+                    ref = guard.obj_weakref
+                    assert ref is not None, "TENSOR_MATCH guard object must be valid"
+                    underlying_tensor_id = id(ref())
+                    if underlying_tensor_id in id_to_arg_mapping.keys():
+                        arg_for_guard_position = id_to_arg_mapping[
+                            underlying_tensor_id
+                        ][0]
+                        arg_for_guard = id_to_arg_mapping[underlying_tensor_id][1]
+
+                        dynamic_dims = None
+                        enforce_shape = None
+                        constrain = False
+
+                        if not torchdynamo.config.dynamic_shapes:
+                            if isinstance(arg_for_guard, PartiallyDynamicTensor):
+                                shape_and_dd = arg_for_guard.shape_and_dynamic_dims()
+                                enforce_shape = shape_and_dd[0]
+                                dynamic_dims = shape_and_dd[1]
+                            else:
+                                enforce_shape = arg_for_guard.shape
+                        else:
+                            if isinstance(arg_for_guard, PartiallyDynamicTensor):
+                                constrain = True
+                                shape_and_dd = arg_for_guard.shape_and_dynamic_dims()
+                                enforce_shape = shape_and_dd[0]
+                                dynamic_dims = shape_and_dd[1]
```

Comment:
Alternative -> Get this from the guard object. Still not sure which I prefer.

Reply:
This is how it would look plumbed into the guard infra.

Not sure that's a valuable direction yet - perhaps if we wanted to move this concept into general support, beyond `export` - https://github.com/pytorch/torchdynamo/commit/b2fa6c97f464737c3a72415ae01931df390f5f7e

Reply:
Collapsed into this one.